### PR TITLE
python lib version with low memory optimizations

### PR DESCRIPTION
**Pull request by olgen**
branched off `remove-invalid-users-low-memory`
- remove all invalid users corrections (racy bug is fixed)
- extract all logic to `/lib/helpers.py`
- adopt for usage in Google Colab + Jupyter environment
- minor tweaks + cleanup

### CODE REVIEW

File: lib/helpers.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
+# -*- coding: utf-8 -*-
+"""## Import needed packages
```
Comment (akaspin) [inline lib/helpers.py]: The file docstring is little confusing. Because it's just various helpers it can be removed.

File: lib/helpers.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+import re
+import os
+import gzip
+import scipy
+import scipy.stats
+import s3fs
+from google.colab import files
+from IPython.display import display  # so we can run this as script as well
+
+"""## Data loading helpers
```
Comment (akaspin) [inline lib/helpers.py]: This is actual file docstring.

File: lib/helpers.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+Define a few helper functions to load and cache data.
+"""
+
+VERSION = '5.4.13'
+cache_folder = "cache-v{0}".format(VERSION)
+
+# constants for groups
```
Comment (akaspin) [inline lib/helpers.py]: This comment is related to only first (`TEST`) constant. The usual ways to document variables are:

"after" doc like:

```
my_var = True
"""This is my variable"""
```

Doc comment like:
```
#: This is my variable
my_var = True
```

or 

```
my_var = True #: This is my variable
```

But anyway each docstring is related to only one variable

File: lib/helpers.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+TEST = True
+CONTROL = False
+
+# use more memory efficient types for ts,user_id and ab_test_group
+def improve_types(df):
```
Comment (akaspin) [inline lib/helpers.py:14]: Usual way to add docstring to functions is "afterdoc" like:

```
def improve_types(df):
    """
    use more memory efficient types for ts,user_id and ab_test_group
    """
    ...
```

File: lib/helpers.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+# column afterwards 
+def extract_revenue_events(df, revenue_event):
+    df = df[df.partner_event == revenue_event]
+    return df.drop(columns=['partner_event'])
+    
+# parquet save and load helper
+def to_parquet(df, filename):
+    df.to_parquet(filename, engine='pyarrow')
+
+# a little "hack" to convert old file on the fly
```
Comment (akaspin) [inline lib/helpers.py]: Maybe more descriptive docstrings is good idea? A description of the behavior of the function and its parameters would be very useful.

File: lib/helpers.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+    print(now, 'start loading CSV for', audience, source, date)
+
+    read_csv_kwargs = {'chunksize': chunk_size}
+    if columns:
+        read_csv_kwargs['usecols'] = columns
+        
+    df = pd.DataFrame()
+    
+    if not fs.exists(path=filename):
+       print(now, 'WARNING: no CSV file at for: ', audience, source, date, ', skipping the file: ', filename)
```
Comment (akaspin) [inline lib/helpers.py]: Maybe logging is more suitable here?

File: lib/helpers.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+            group_bids_df = bids_df[bids_df.campaign_id.isin(campaigns)]
+            report_df[name] = uplift(group_bids_df, attributions_df, name, use_converters_for_significance, len(groups))
+            
+    if report_df is not None and per_campaign_results:
+        campaigns = bids_df['campaign_name'].unique()
+        for campaign in campaigns:
+            name = "c_{0}".format(campaign)
+            campaign_bids_df = bids_df[bids_df.campaign_name == campaign]
+            report_df[name] = uplift(campaign_bids_df, attributions_df, name, use_converters_for_significance, len(campaigns))
+    return report_df
```
Comment (akaspin) [inline lib/helpers.py]: endline?

File: lib/helpers.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    return pd.DataFrame(
+        dataframe_dict,
+        index=[index_name],
+    ).transpose()
+
+"""### Calculate and display uplift report for the data set as a whole
+This takes the whole data set and calculates uplift KPIs.
+"""
+
+def uplift_report(bids_df, attributions_df, groups, per_campaign_results, use_converters_for_significance):
```
Comment (akaspin) [inline lib/helpers.py:383]: (low) Both `uplift` and `uplift_report` just ask to be converted into one class.

File: lib/helpers.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+        
+    df = pd.DataFrame()
+    
+    if not fs.exists(path=filename):
+       print(now, 'WARNING: no CSV file at for: ', audience, source, date, ', skipping the file: ', filename)
+       return df
+      
+    for chunk in pd.read_csv(filename, escapechar='\\', low_memory=False,
+                             **read_csv_kwargs):
+        if chunk_filter_fn:
```
Comment (akaspin) [inline lib/helpers.py:139]:
```
filtered_chunk = chunk_filter_fn(chunk, revenue_event) if chunk_filter_fn else chunk
```

File: lib/helpers.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+
+        filtered_chunk = improve_types(filtered_chunk)
+
+        df = pd.concat([df, filtered_chunk],
+                       ignore_index=True, verify_integrity=True)
+
+    print(datetime.now(), 'finished loading CSV for', date.strftime('%d.%m.%Y'),
+          'took', datetime.now() - now)
+
+    if not os.path.exists(cache_dir):
```
Comment (akaspin) [inline lib/helpers.py:157]:
```
os.makedirs(cache_dir, exist_ok=True)
```

**Review by akaspin — CHANGES_REQUESTED** [review]
Looking functional. But docs and comments are very confusing on first sight.

**Review by akaspin — APPROVED** [review]

File: lib/helpers.py (hunk 6 above)
Comment (VOvchinnikov) [reply lib/helpers.py]: Maybe later, when there's a class wrapping this thing up

File: lib/helpers.py (hunk 8 above)
Comment (VOvchinnikov) [reply lib/helpers.py:383]: ...and a bunch of other things :)

**Pull request merged**